### PR TITLE
Average Precision to OD Utils

### PR DESCRIPTION
**Pull request by marko-polo-cheno**
Average Precision calculation

### CODE REVIEW

File: kolena/_experimental/object_detection/utils.py [hunk 1]
```diff
@@ -267,3 +267,40 @@ def compute_optimal_f1_threshold_multiclass(
     for label, (y_true, y_score) in y_true_score_by_label.items():
         optimal_thresholds[label] = _compute_optimal_f1_with_arrays(y_true, y_score)
     return optimal_thresholds
+
+
+def compute_ap(precisions: List[float], recalls: List[float]) -> float:
```
Comment (gordonhart) [inline kolena/_experimental/object_detection/utils.py]: Nit: `compute_average_precision`

File: kolena/_experimental/object_detection/utils.py [hunk 2]
```diff
@@ -267,3 +267,40 @@ def compute_optimal_f1_threshold_multiclass(
...
+    """
+    Computes the average precision given a PR curve with the metrics methodology of Pascal VOC.
+    Based on https://github.com/Cartucho/mAP which implements the Matlab Pascal Voc code in Python.
```
Comment (gordonhart) [inline kolena/_experimental/object_detection/utils.py]: Formatting can be improved — these docstrings are now just Markdown. Also ensure that capitalization is consistent, [elsewhere](https://github.com/kolenaIO/kolena/blob/trunk/kolena/workflow/metrics/_geometry.py#L184) we refer to this as `PASCAL VOC`.

File: kolena/_experimental/object_detection/utils.py [hunk 3]
```diff
@@ -267,3 +267,40 @@ def compute_optimal_f1_threshold_multiclass(
...
+    :param recalls: A list recall values from a PR curve.
+    :return: The value of the average precision.
+    """
+
+    if len(precisions) != len(recalls):
+        raise ValueError("precisions and recalls differ in length")
+
+    pairs = sorted(zip(recalls, precisions), key=lambda x: x[0])
+    recalls = [x[0] for x in pairs]
+    precisions = [x[1] for x in pairs]
```
Comment (gordonhart) [inline kolena/_experimental/object_detection/utils.py]: This isn't important, but for readability it's a good practice to use destructuring where possible. Here that would mean:

```python
recalls = [recall for recall, _ in pairs]
precision = [precision for _, precision in pairs]
```

**Review by gordonhart — APPROVED** [review]

File: tests/unit/_experimental/test_utils.py [hunk 4]
```diff
@@ -266,3 +266,81 @@ def test__metrics__f1__optimal__multiclass(
...
+        ([1, 1, 0.99, 0.95, 0.90, 0.85, 0.8], [0.03, 0.10, 0.356, 0.55, 0.65, 0.7, 0.74], 0.70224),
+        # straight chunk + curve
+        ([1, 1, 1, 0.992, 0.973, 0.9, 0.84], [0.002, 0.1, 0.33, 0.4, 0.5, 0.6, 0.63], 0.61194),
+        # another curve with precision rising and falling, but defined over a small recall
+        (
+            [1, 1, 0.88, 0.75, 0.82, 0.8, 0.75, 0.7],
+            [0.003, 0.02, 0.02, 0.03, 0.05, 0.08, 0.08, 0.11],
+            0.08959999999999999,
+        ),
+        # zero precisions
```
Comment (y27choi) [inline tests/unit/_experimental/test_utils.py:305]: Let's add zero recalls case and both precisions and recalls zero case as well.

**Review by y27choi — APPROVED** [review]

**Pull request merged**